### DISCUSSION TITLE
[Typing] - `HTTPError` doesn't have a `request` attribute

### DISCUSSION THREAD

**Discussion by michaeloliverx** [encode/httpx#2151] [Potential Issue] [answered]
The `HTTPError` class serves as a base class for `RequestError` and `HTTPStatusError`. The docstring shows the following usage:
```python
try:
    response = httpx.get("https://www.example.com")
    response.raise_for_status()
except httpx.HTTPError as exc:
    print(f"HTTP Exception for {exc.request.url} - {exc}")
```

Mypy doesn't like this:
```
error: "HTTPError" has no attribute "request"  [attr-defined]
```

**Comment by florimondmanca** [comment]
Both child classes do have `.request` but it does seem like mypy isn't able to tell that. Should this be treated as an HTTPX issue (should we be adding `request: Request` annotation to `HTTPError` as a hint), or is this something mypy should be able to infer? Is there any open mypy ticket on a case like this?

**Reply by michaeloliverx** [reply]
The following change keeps mypy happy in my test:
```diff
class HTTPError(Exception):
    """
    Base class for `RequestError` and `HTTPStatusError`.

    Useful for `try...except` blocks when issuing a request,
    and then calling `.raise_for_status()`.

    For example:

    ```
    try:
        response = httpx.get("https://www.example.com")
        response.raise_for_status()
    except httpx.HTTPError as exc:
        print(f"HTTP Exception for {exc.request.url} - {exc}")
    ```
    """

+   request: "Request"

    def __init__(self, message: str) -> None:
        super().__init__(message)
```


One issue with this approach is that we are essentially lying to the type checker that `HTTPError` has a `request` attribute defined when in fact it is only the subclasses that define it. We rely on all subclasses to define that attribute (which they do atm). Perhaps it should be a real attribute on `HTTPError` rather than a hint?

**Reply by michaeloliverx** [reply]
There is no way mypy would be able to infer an attribute that isn't declared.

**Reply by florimondmanca** [reply]
> There is no way mypy would be able to infer an attribute that isn't declared.

Well, true.

Maybe we could do this…

```python
class HTTPError(Exception):
    def __init__(self, message: str) -> None:
        super().__init__(message)
        self._request: Optional["Request"] = None

    @property
    def request(self) -> "Request":
        if self._request is None:
            raise RuntimeError("The .request property has not been set.")
        return self._request


class RequestError(HTTPError):
    def __init__(self, message: str, *, request: "Request" = None) -> None:
        super().__init__(message)
        self._request = request

    @HTTPError.request.setter  # XXX: Does this even work???
    def request(self, request: "Request") -> None:
        self._request = request


class HTTPStatusError(HTTPError):
    def __init__(
        self, message: str, *, request: "Request", response: "Response"
    ) -> None:
        super().__init__(message)
        self._request = request
        self.response = response
```

Btw that `setter` on `RequestError` is used to attach the request during a request/response cycle: https://github.com/encode/httpx/blob/317653585cd781c59fc3f5dde20292a47f765a0e/httpx/_exceptions.py#L336-L339

Actually currently `HTTPStatusError` defines `.request` as a public (read-write) attribute so the setter could very much be defined in `HTTPError`…

**Reply by michaeloliverx** [reply]
I think putting the property getter/setter on `HTTPError` makes sense.

**Comment by michaeloliverx** [comment, marked as answer]
Fixed in https://github.com/encode/httpx/pull/2158
